torch/csrc/autograd/python_function.cpp

```diff
@@ -1079,8 +1086,12 @@ bool THPFunction_initModule(PyObject *module)
 
 struct Decref {
   void operator()(PyFunction* p) const {
-    AutoGIL gil;
-    Py_DECREF(p->obj);
+    if (swappedThreadState) {
```

Comment:
Wouldn't it be less error prone if you just had `AutoGIL` take care of deciding whether or not the GIL needs to be taken out again? Also, the Python invariants for when you're allowed to take out a GIL while you have it, and when you are not, deserve a beefy note somewhere in the code :)

Reply:
Yeah, you've edited a bunch of AutoGIL sites, and I don't have very much confidence that you've necessarily gotten all of them :)

Alternately, could we perhaps make it a BUG to take out AutoGIL while you hold it, and then fix up cases where we made this mistake? (I could believe that the sites you've identified are exactly those). Something like a `TORCH_INTERNAL_ASSERT` in AutoGIL that we don't already have the GIL. This would be a reasonable patch which we could land before this one.